### ISSUE TITLE
Anybody else seeing random read-corruption on Amiga .ADF files?

### ISSUE THREAD

**Issue by rjflory** [keirf/flashfloppy#748]
Gotek HW: very-recent Artery AT32F435RG17 SoC.
Gotek FW: 3.37

Target: Two different Amiga A500 systems w/ WB 1.3 .
 
 I have two different Gotek units.  Since I got these a few weeks ago, I've seen random read-corruption at random times at random places.  This has been the source of many hours of debug 'amusement.'..  ;)
 
 I hope this information will be helpful in the debug process.

 - Sometimes programs or text files will be reported as corrupt or will crash the target A500 system(s).
 - Sometimes the A500 target system will ask me to re-insert the Gotek boot-floppy when it has not been removed, written, etc.
 - In one experiment, using multiple A500 targets, I 'cloned' the Gotek boot disk (Amiga Workbench 1.3, disk 1) onto four different 'empty' files of a second gotek drive, running the same FW as the first.  
   - All of the clone operations performed successful read-after-write verifies.
   - Two of the 'clone' files had just one corrupt file in its .adf filesystem.  
   - The directory entries and file-lengths were correct.
   - The corrupt regions were exactly 64-bytes in length in both of the "BAD" cases.  Data following the corruption was OK.
   - The corrupt regions were at different offsets into the file(s), seemingly aligned on 8-byte boundary.
   - In BOTH cases, the corrupt region was a duplication of the preceding 64-bytes of the file. 
   - I tested this using two different Amiga A500 systems (one a Rev-5, the other a Rev-6A).
   - One corrupt file (l/Aux-Handler) was 2400-bytles long, the other (l/FastFileSystem) was 9388-bytes long. 
  
 - The first (boot/read) USB-flash was a (physically) tiny 32GB Sandisk SDCZ43 USB 3.0 device with a 8GB FAT32 first-partition, the second (written) flash was a Sandisk SDCZ50-008G Cruzer Blade 8GB.
 - I'm using x86 Fedora Linux to access the .adf images as loopback mounted AFFS filesystems.
 
ron
embedded-systems software engineer-

![Screenshot_2023-01-22_15-48-50_cr](https://user-images.githubusercontent.com/123330571/213942362-02633fc7-cad9-4f00-b0a6-2dbd47e0960b.png)

**Comment by keirf** [comment]
Thanks for the report! What was the method used for cloning? Was the read and write done on the same Gotek?

**Comment by rjflory** [comment]
Amiga gui-based "DiskCopy" (basically drag the "WorkBench" boot disk icon onto the "Empty" disk icon).  The operation if read-after-write verify on a per-track basis.

**Comment by keirf** [comment]
Thanks for the report! What was the method used for cloning? Was the read and write done on the same Gotek?

I think perhaps a good test will be to hook up to a Greaseweazle and then loop: create a random adf, write to Gotek, read from Gotek, check the random data. That would be close enough, and should detect badness committed to USB drive as there is not enough sram to cache a whole disk. Whereas track by track verify will hit the cache.

**Comment by keirf** [comment]
415 has so much testing in the field with no reports, I feel this is probably a bug specific to 435.

**Comment by rjflory** [comment]
I added a side-by-side hexdump to original port.
good on left,  bad on right.

**Comment by rjflory** [comment]
??? 435 ???   do you mean 4.35?

 is this an internal number.  FW version 3.37 is mentioned on original post.

**Comment by rjflory** [comment]
I guest I should try a beta version--  do you recommend a "trusted" (maybe RC) version?

**Comment by keirf** [comment]
Thanks. Looks like 64 bytes bad? Interesting that it doesn't seem to be naturally aligned. 

Duplicate of previous 64 bytes... I wonder if it's a USB issue since packet size is 64 bytes.

**Comment by rjflory** [comment]
duh...   4 lines of 16 bytes is 64, not 32...

**Comment by rjflory** [comment]
yes, in both cases, the miscompare was aligned to xx80 offset in the file.

**Comment by rjflory** [comment]
Yes, the possibility of flakey USB drive has been a suspect.  Will have to try some alternate flash drives as well as alternate Gotek FW to narrow down further.

**Comment by keirf** [comment]
I'd have to go back over how USB MSC ops get stuffed into packets. There could be an eight-byte header leading straight into data for example. 

Regarding "435", I'm talking about the MCU model AT32F435. The more common model currently is AT32F415 which runs a different firmware build. Did you unexpectedly receive AT32F435? The factory did say that at some point they might switch over to this new more capable chip for their mass production. https://github.com/keirf/flashfloppy/wiki/Gotek-Models#microcontroller-options

**Comment by keirf** [comment]
Trying the latest v4.x beta firmware could be interesting as the I/O path is quite different. Although the USB stack is the same.

I wouldn't worry about trying this if it is much bother.

**Comment by rjflory** [comment]
In regard to trying 4.x vs 3.x code, are there any known issues regarding upgrade vs downgrade compatability?  I'd hate to brick one...

The person selling these Goteks on ebay touted "uses latest version cpu", so yes, I was aware of some aspect of this and the recent switch-over to new Soc.    ;)

**Comment by rjflory** [comment]
BTW, I have noticed that the verify-portion of the Amiga DiskCopy operation will fail at the same place if some slower USB 2.0 flash drives are used.  This would suggest for **this** device, the cache was not hit for the verify.

**Comment by keirf** [comment]
You can upgrade and downgrade firmware easily. It's also actually easy to entirely reflash with a usb-a to a or c cable and a Windows PC.

**Comment by keirf** [comment]
The problem of failed verify is probably timeout. Currently FlashFloppy will write back to USB before starting to serve read data again. If the write back is slow this could be a problem.

In this regard the v4.x beta firmware is interesting as it writes back only when cylinder changes (when drive heads are stepped). Hence serves reads after writes with much lower and predictable latency.

**Comment by rjflory** [comment]
I have upgraded the firmware to 3.37 already, was just asking if there were known incompatibility issues when downgrading from 4.x to earlier 3.x .

**Comment by keirf** [comment]
Downgrade will work fine. However you can't go earlier than about 3.34 as that's when support for 435 was added

**Comment by rjflory** [comment]
OK, thats exactly the sort of compatibility info I was looking for.

btw, is 4.4a the latest,  links for 4.6a do not appear to be 'live'? (or ready yet)

**Comment by keirf** [comment]
4.7a is latest. Link is at top of downloads page on the wiki.

4.6a link does appear to be broken. I will fix it but it's not the latest anyway.

**Comment by keirf** [comment]
Just an update: I have been writing ADF images as fast as possible with no verify and no sync to index using Greaseweazle:
```
dd if=/dev/urandom of=a.adf bs=1024 count=880
gw write a.adf --fake-index=300 --no-verify
gw read b.adf
diff a.adf b.adf
```
I haven't found any corruptions so far. This is FF `stable-v3` tip (v3.38 plus a few changes on top) running on a AT32F435 Gotek.

I will say that v3.38 has a 435 specific fix since v3.37. I'm not sure it would make a difference in your setup but it's a possibility.

**Comment by rjflory** [comment]
Some thoughts.
- It would be nice to run that verification test over multiple iterations.
- That particular test writes/reads the data as a sequential operation (the least stressful of all scenarios), I wonder if interleaved random seeks to non-sequential tracks/sectors (as one would experience due to target directory and filesystem access) might aggravate problems with the SoC USB-driver, or within the flash drive itself. 
 - Can Greaseweasle do random seeks and read/write 'n' sector-length blocks?  This would allow simple, yet very thorough tests to be scripted and run over extended periods of time.  A nice regression/verification suite...
 
 btw, I just ordered two Greaseweazle units from a Canadian source.  They seem like very handy devices for those of us who occasionally work with, and develop for vintage equipment.  

 It will be nice to have an alternate means of testing/verification the Gotek & HXC drive emulators.

**Comment by keirf** [comment]
Random seek would be a good stress test, however I focused on errors introduced by diskcopy, which is a sequential operation (except for disk swaps). Greaseweazle does whole track reads and writes. But then so does Amiga, even when accessing single sectors.

I do have another project called flashfloppy-testbed which allows a modified Gotek to test another, and this can do random (IBM format) sector accesses and the like. It gets run fairly routinely. https://github.com/keirf/flashfloppy-testbed

Other debug aids include a debug build of FlashFloppy which emits logs at 3Mbaud on the serial pins at the programming header at the rear of the Gotek. I can use this to observe write latencies to USB, for example.

**Comment by rjflory** [comment]
OK, good information.

**Comment by keirf** [comment]
Perhaps a good test would be random track reads from a pseudorandom ADF in the testbed. Then independently verify the sector contents based on knowledge of the PRNG seed.

**Comment by rjflory** [comment]
I got distracted by other things, but should have left a note.

 Reflashing my Goteks to 4.7a firmware seems to have resolved the write-corruption issues I was seeing.  I've done many diskcopies and other write ops since, and have not seen this problem since.

**Comment by keirf** [comment]
Thanks for the follow up!
